### PR TITLE
fix: ensure CSS URLs are correct when inlining client stylesheets

### PR DESCRIPTION
**Pull request by teemingc**
fixes https://github.com/sveltejs/kit/issues/13878
closes https://github.com/sveltejs/kit/pull/15147

This PR replaces the janky server-client css mapping (introduced by me) with simply replacing the asset URL references with the correct path when we inline the styles. Not only does this massively simplify things, it allows us to correctly prefix the paths based on the settings from `paths.relative` or `path.assets`.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

File: packages/kit/test/apps/options/package.json [hunk 1]
```diff
@@ -9,8 +9,8 @@
 		"prepare": "svelte-kit sync",
 		"check": "svelte-kit sync && tsc && svelte-check",
 		"test": "pnpm test:dev && pnpm test:build",
-		"test:dev": "DEV=true playwright test",
-		"test:build": "playwright test",
+		"test:dev": "DEV=true playwright test && DEV=true PATHS_ASSETS=https://cdn.example.com/stuff playwright test",
+		"test:build": "playwright test && PATHS_ASSETS=https://cdn.example.com/stuff playwright test",
```
Comment (teemingc) [inline packages/kit/test/apps/options/package.json:13]: This seemed like the easiest way to test both CSS inlining with and without paths.assets without introducing another test matrix.

paths.assets replaces the CSS asset urls at build time but paths.relative does it at runtime

File: packages/kit/src/utils/css.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Vite's static asset handling for the client changes the asset URLs in a CSS
+ * file to start with `./`. This is incorrect if we're inlining the CSS or if
+ * `paths.assets` is set. This function helps us rewrite the URLs in the CSS.
+ * @param {string} contents
+ * @param {string} base
+ * @returns {string}
+ */
+export function replace_css_relative_url(contents, base) {
+	return contents.replaceAll(/url\((['"]?)\.\//g, `url($1${base}/`);
```
Comment (teemingc) [inline packages/kit/src/utils/css.js]:
```suggestion
	return contents.replaceAll(/url\((['"]?)\.\//ig, `url($1${base}/`);
```

File: packages/kit/src/utils/css.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Vite's static asset handling for the client changes the asset URLs in a CSS
+ * file to start with `./`. This is incorrect if we're inlining the CSS or if
+ * `paths.assets` is set. This function helps us rewrite the URLs in the CSS.
+ * @param {string} contents
+ * @param {string} base
+ * @returns {string}
+ */
+export function replace_css_relative_url(contents, base) {
+	return contents.replaceAll(/url\((['"]?)\.\//ig, `url($1${base}/`);
```
Comment (teemingc) [inline packages/kit/src/utils/css.js]: Fortunately, we don't have to worry about the `@import 'app.css'` case that omits the `url(...)` syntax because Vite inlines the referenced stylesheet even when referenced by more than one other stylesheet https://vite.dev/guide/features#import-inlining-and-rebasing

File: packages/kit/src/utils/css.spec.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		);
+	});
+
+	test('handles double-quoted urls', () => {
+		assert.equal(
+			replace_css_relative_url('url("./image.png")', '/assets'),
+			'url("/assets/image.png")'
+		);
+	});
+
```
Comment (teemingc) [inline packages/kit/src/utils/css.spec.js:206]:
```suggestion

	test('case-insensitive urls', () => {
		assert.equal(
			replace_css_relative_url('uRl("./image.png")', '/assets'),
			'url("/assets/image.png")'
		);
	});

```

File: packages/kit/src/utils/css.js (hunk 3 above)
Comment (teemingc) [inline packages/kit/src/utils/css.js]:
```suggestion
	return contents.replaceAll(/url\((['"]?)\.\//gi, `url($1${base}/`);
```

**Comment by teemingc** [comment]
> Is it a problem that this doesn't work on relative paths without ./ in front of them? i.e. url(img.png) won't get transformed.

I think after Vite processes the CSS the urls are always prefixed with ./ but I can check how it’s handled from both the static folder or otherwise to see if they do end up this way.

> This will replace content inside strings. So if you had the CSS content: "Example: If you were to put url(./rickroll.gif) here...", it would get modified, in spite of being a string. I don't see a way around this other than actually parsing the CSS contents with a CSS parser...

Yeah, I think I should try out a CSS parser. Otherwise, maybe we’d have to do a negative lookahead. Are there any CSS parsers that you’d recommend?

EDIT: I see we were previously using csstree in Svelte so we can probably try that?

**Comment by teemingc** [comment]
> * Is it a problem that this doesn't work on relative paths without `./` in front of them? i.e. `url(img.png)` won't get transformed.

So I did some testing and here are the results:
1. If the asset can be found, it's always suffixed with `./` and lives next to the CSS file.
2. If the asset can be found and it's located in the `static` folder, it's suffixed with `../../../` because static assets are located at the root path and it's relative to `_app/immutable/assets`.
3. If the asset can't be found, the url value remains the same.

**Comment by teemingc** [comment]
- [x] We'll use the Svelte CSS parser to identify declaration values and avoid accidentally affecting comments, strings, etc.
- [x] We'll also transform the stored CSS from a string to a function so we can easily plug in the relative path at runtime
- [x] We'll need to handle the case where the asset path is referencing something from the root because it was placed in a static dir. I've already added a test for this.

File: packages/kit/src/exports/vite/build/build_server.js [hunk 5]
```diff
@@ -76,6 +86,30 @@ export async function build_server_nodes(
 		static_exports
 	});
 
+	/**
+	 * For CSS inlining, we either store a string or a function that returns the
+	 * styles with the correct asset base path
+	 * @type {(css: string, eager_assets: Set<string>) => string}
+	 */
+	let prepare_css_for_inlining = (css) => s(css);
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/exports/vite/build/build_server.js]: Can't this just be this:


```suggestion
	let prepare_css_for_inlining = s;
```

File: packages/kit/src/exports/vite/build/build_server.js [hunk 6]
```diff
@@ -180,19 +215,25 @@ export async function build_server_nodes(
 			`export const fonts = ${s(fonts)};`
 		);
 
+		if (eager_assets.size) {
+			console.log(eager_assets);
+			// TODO: strip immutable path prefix from assets?
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/exports/vite/build/build_server.js]: todo

File: packages/kit/src/utils/css.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+/** @typedef {{ property: string; value: string; start: number; end: number; type: 'Declaration' }} Declaration */
+
+/**
+ * Vite's static asset handling for the client changes the asset URLs in a CSS
+ * file to start with `./` or `../../../`. This is incorrect if we're inlining
+ * the CSS or if `paths.assets` is set, so we need to fix them.
+ * @param {string} css
+ * @param {Set<string>} known_assets
+ * @param {string} assets
+ * @param {string=} base
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: Is this yet another syntax for an optional parameter? 😆 Can we do the `[base]` version instead?

File: packages/kit/src/utils/css.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+			if (!known_assets.has(filename)) return;
+
+			if (prefix === './') {
+				declaration.value = `url(${quote}${assets}/${filename}${quote})`;
+			} else {
+				declaration.value = `url(${quote}${base}/${filename}${quote})`;
+			}
+
+			s.update(declaration.start, declaration.end, `${declaration.property}:${declaration.value};`);
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: If we're using `svelte.parse` shouldn't we be adding an offset to this? becuase of the additional `<style>` at the beginning

File: packages/kit/src/exports/vite/build/build_server.js (hunk 5 above)
Comment (teemingc) [reply packages/kit/src/exports/vite/build/build_server.js]: It should but typescript complains about the second argument being passed in as a replacer

File: packages/kit/src/utils/css.js (hunk 7 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: Resolving this because it's no longer optional

File: packages/kit/src/utils/css.js (hunk 8 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: Is it correct to do `declaration.start - '<style>'.length` if we are using svelte.parse?

Comment (elliott-with-the-longest-name-on-github) [reply packages/kit/src/utils/css.js]: Yeah I think so, though we would have to do the same to the `end` as well. Gonna have to test to make sure

**Comment by teemingc** [comment]
Ok, I think it's ready. Some additional edge cases accounted for are URLs with escaped quotes, fragments, query strings, etc.. Only known assets are processed too so it doesn't just match anything with `./` or `../../../`.

File: packages/kit/src/utils/css.js (hunk 8 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: I tested this briefly just by removing the `parseCss` logic and running the tests.

File: packages/kit/test/apps/options/test/test.js
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/test/apps/options/test/test.js:1]: Is it possible to just split the `PATHS_ASSETS-affected` tests into their own file and then skip this whole file when we need to? I mainly worry we'll start polluting the second test run the first time we add a `test.describe` block and forget to do the right thing with `test.skip`.

File: packages/kit/src/utils/css.spec.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+				vite_assets: new Set(['bg.png', 'border.svg', 'mask.png']),
+				static_assets: new Set(),
+				assets: './_app/immutable/assets',
+				base: '.'
+			}),
+			expected
+		);
+	});
+
+	test('does not replace non-relative urls', () => {
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.spec.js]: This and the next one might be better-served with `test.each` blocks

File: packages/kit/src/utils/css.spec.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+				vite_assets: new Set(['a.png', 'b.png']),
+				static_assets: new Set(),
+				assets: './_app/immutable/assets',
+				base: '.'
+			}),
+			expected
+		);
+	});
+
+	test('does not replace relative url without ./ prefix', () => {
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.spec.js]: Good coverage, I might just look at using a `test.each` block with an array of cases instead of having to repeat `fix_css_urls` with the exact same `vite_assets`/`static_assets`/`assets`/`base` over and over

File: packages/kit/src/utils/css.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
+import MagicString from 'magic-string';
+import * as svelte from 'svelte/compiler';
+
+/** @typedef {ReturnType<typeof import('svelte/compiler').parseCss>['children']} StyleSheetChildren */
+
+/** @typedef {{ property: string; value: string; start: number; end: number; type: 'Declaration' }} Declaration */
+
+const parser = svelte.parseCss
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: nit, but this should probably be named `parse`

File: packages/kit/src/utils/css.js [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+const VITE_ASSET_PREFIX = './';
+
+const STATIC_ASSET_PREFIX = '../../../';
+
+const FRAGMENT_OR_QUERY_REGEX = /[?#]/;
+
+/**
+ * Vite's static asset handling for the client changes the asset URLs in a CSS
+ * file to start with `./` or `../../../`. This is incorrect if we're inlining
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: Is it guaranteed to be exactly `../../../` always, regardless of user config? I'm trying to think of any possible edge cases with monorepo config...

File: packages/kit/src/utils/css.js [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+ * the CSS or if `paths.assets` is set, so we need to fix them.
+ * @param {{
+ * 	css: string;
+ * 	vite_assets: Set<string>;
+ * 	static_assets: Set<string>;
+ * 	assets: string;
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: This should probably be `assets_path` or something (open to suggestions) -- it's confusing along with `vite_assets` and `static_assets`

File: packages/kit/src/utils/css.spec.js
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.spec.js:1]: We should also have a test making sure it doesn't replace content inside of comments / in unrelated strings like `content: "url('./image.png')"`

File: packages/kit/src/utils/css.js [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	const parsed = parser(css);
+
+	for (const child of parsed.children) {
+		find_declarations(child, (declaration) => {
+			/** @type {string} */
+			let new_value = declaration.value;
+
+			/** @type {RegExpExecArray | null} */
+			let url_declaration_match;
+			const url_declaration_regex = /url\(\s*[^)]*\)/gi;
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: This should be moved to a top-level const so we don't have to recreate it on every callback

File: packages/kit/src/utils/css.js [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+			while ((url_declaration_match = url_declaration_regex.exec(declaration.value))) {
+				const [url_declaration] = url_declaration_match;
+
+				const url_value_match = /url\(\s*(['"]?)(.*?)\1\s*\)/i.exec(url_declaration);
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: same for this one

File: packages/kit/src/utils/css.js [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+				if (!url_value_match) continue;
+
+				const [, , url] = url_value_match;
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: I think you could use non-capturing groups (`(?:)`) for the stuff you don't actually want to capture to avoid these weird commas

File: packages/kit/src/utils/css.js [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+					[filename] = url.slice(STATIC_ASSET_PREFIX.length).split(FRAGMENT_OR_QUERY_REGEX);
+
+					if (current_prefix === STATIC_ASSET_PREFIX && static_assets.has(filename)) {
+						new_prefix = base;
+					}
+				}
+
+				if (!new_prefix) continue;
+
+				new_value = new_value.replace(`${current_prefix}${filename}`, `${new_prefix}/${filename}`);
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js:126]: Do `assets` and `base` ever end with a slash? Just checking, as this would cause it to have a double slash

File: packages/kit/src/utils/css.js
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js:1]: I did an AI-assisted analysis after my first pass (all of the comments below come from me), but the AI seems like it might have caught a few issues:


# Analysis of Potential Bugs in `packages/kit/src/utils/css.js`

Based on PR #15153 (which fixes CSS URLs when inlining client stylesheets) and the associated issue #13878 (CSS styles not applied with non-zero inlineStyleThreshold).

## 1. `String.prototype.replace()` only replaces the first occurrence (line 86)

```javascript
new_value = new_value.replace(`${current_prefix}${filename}`, `${new_prefix}/${filename}`);
```

This uses `.replace()` with a plain string, which **only replaces the first occurrence**. If the same URL appears multiple times in a single declaration value (e.g., in a gradient or `image-set()`), subsequent occurrences won't be fixed.

**Example that would fail:**

```css
div {
	background: image-set(url(./a.png) 1x, url(./a.png) 2x);
}
```

The second `./a.png` would remain unchanged.

---

## 2. Potential double-slash in replaced URL (line 86)

```javascript
new_value = new_value.replace(`${current_prefix}${filename}`, `${new_prefix}/${filename}`);
```

If `new_prefix` already ends with a `/` (e.g., `assets = "https://cdn.example.com/"`), the replacement would produce `https://cdn.example.com//filename.png` with a double slash.

The test cases use values like `'https://cdn.example.com/_app/immutable/assets'` (no trailing slash), but there's no validation that the caller won't pass a trailing slash.

---

## 3. URL regex may not handle all edge cases correctly (line 59)

```javascript
const url_value_match = /url\(\s*(['"]?)(.*?)\1\s*\)/i.exec(url_declaration);
```

The `(.*?)` is non-greedy and matches "anything", but it might incorrectly capture content in edge cases:

- URLs with escaped parentheses
- URLs with escaped quotes matching the outer quote character

For example, `url("image\")other.png")` could potentially be mis-parsed.

---

## 4. The outer URL regex stops at the first `)` (line 55)

```javascript
const url_declaration_regex = /url\(\s*[^)]*\)/gi;
```

The `[^)]*` will stop at the first `)` it encounters. This could fail with URLs containing encoded parentheses. More complex nesting could also be problematic:

```css
div {
	background: calc(url(./img.png));
}
```

The regex would match `url(./img.png)` correctly here, but edge cases with encoded `)` characters in URLs could cause issues.

---

## 5. Filename extraction doesn't account for URL-encoded characters

```javascript
let [filename] = url.slice(VITE_ASSET_PREFIX.length).split(FRAGMENT_OR_QUERY_REGEX);
```

If a filename contains URL-encoded characters (e.g., `image%20name.png`), it would be compared against the `vite_assets` Set which might contain the decoded version (`image name.png`), causing a mismatch and the URL not being fixed.

---

## Summary

The most critical bug is **#1** - using `.replace()` instead of `.replaceAll()` means duplicate URLs in the same declaration value won't all be fixed. This seems like an oversight given that the loop handles multiple `url()` occurrences across different declarations but not multiple occurrences of the same asset URL within a single declaration value.

File: packages/kit/test/apps/options/test/test.js
Comment (teemingc) [reply packages/kit/test/apps/options/test/test.js:1]: Good idea. This is so much better. I've moved them into `paths-assets.test.js`

File: packages/kit/src/utils/css.js (hunk 14 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: I thought I had to redeclare it each time to correctly continue executing the regex on the same string. Turns out, we can hoist it but reset the `.lastIndex` property when we start on a new string. Thanks!

File: packages/kit/src/utils/css.js (hunk 16 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: The quotation's using a capture group so that it matches against the same type of closing quotation mark. This avoids matching a double quote in a single quoted string.

Unfortunately, this doesn't account for backslashed quotes in the same string

File: packages/kit/src/utils/css.js (hunk 12 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: Gah you're right. ~I knew we controlled where the assets go to (which is always `_app/immutable/assets` unless `paths.assets` specifies a full href), but~ the static folder _can_ be configured via `files.assets` https://svelte.dev/docs/kit/configuration#files . I'll investigate further.

EDIT: the app folder can be changed too 🤔 https://svelte.dev/docs/kit/configuration#appDir I'll test what kind of prefix ends up being appended if both options are configured

EDIT2: Ok so it doesn't matter if `paths.assets` is changed because the files are copied to the root of the build output directory anyway. `appDir` is a different case though. We only have to watch out for deeply nested directories such as `appDir: "_example/nested"`. Vite doesn't allow relative or absolute paths so we have less to worry about:
```
error during build:
Invalid pattern "../../_wheee/immutable/assets/[name].[hash][extname]" for "output.assetFileNames", patterns can be neither absolute nor relative paths. If you want your files to be stored in a subdirectory, write its name without a leading slash like this: subdirectory/pattern.
```

EDIT3: Alright, I've made it calculate the prefix for static assets based on the path where the Vite assets are stored.

File: packages/kit/src/utils/css.spec.js
Comment (teemingc) [reply packages/kit/src/utils/css.spec.js:1]: Good catch! I've added a test for both and they're currently failing.

EDIT: ugh the regex I'm using to strip out strings and comments is too naive and can incorrectly match comment markers or quotes inside strings. We might be better off using something like tippex but for CSS or parsing each character similar to the Svelte CSS parser

EDIT2: I've tried implementing a value parser similar to the Svelte CSS parser but it whites out string and comment contents instead.

File: packages/kit/src/utils/css.js (hunk 17 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js:126]: I don't think they do. At the very least, we're throwing errors if they are ever configured with one https://github.com/sveltejs/kit/blob/fix-missing-css-with-inlining-enabled/packages/kit/src/core/config/index.spec.js#L285 . If `base` does get set to `/`, it probably means `paths.relative` is false and we never end up here anyway.

EDIT: I've added a safeguard to strip the trailing slash just in case

File: packages/kit/src/utils/css.js
Comment (teemingc) [reply packages/kit/src/utils/css.js:1]: > ## 1. `String.prototype.replace()` only replaces the first occurrence (line 86)
> ```js
> new_value = new_value.replace(`${current_prefix}${filename}`, `${new_prefix}/${filename}`);
> ```
> 
> This uses `.replace()` with a plain string, which **only replaces the first occurrence**. If the same URL appears multiple times in a single declaration value (e.g., in a gradient or `image-set()`), subsequent occurrences won't be fixed.
> 
> **Example that would fail:**
> 
> ```css
> div {
> 	background: image-set(url(./a.png) 1x, url(./a.png) 2x);
> }
> ```
> 
> The second `./a.png` would remain unchanged.

This shouldn't be an issue because we're using a separate regex to capture `url(...)` before searching for the value inside it. Then, we move on to the next `url(...)` for the same declaration.

> ## 2. Potential double-slash in replaced URL (line 86)
> ```js
> new_value = new_value.replace(`${current_prefix}${filename}`, `${new_prefix}/${filename}`);
> ```
> 
> If `new_prefix` already ends with a `/` (e.g., `assets = "https://cdn.example.com/"`), the replacement would produce `https://cdn.example.com//filename.png` with a double slash.
> 
> The test cases use values like `'https://cdn.example.com/_app/immutable/assets'` (no trailing slash), but there's no validation that the caller won't pass a trailing slash.

The asset and base paths should never have a trailing slash. We _could_ add a check and remove it if ever does get through.

> ## 3. URL regex may not handle all edge cases correctly (line 59)
> ```js
> const url_value_match = /url\(\s*(['"]?)(.*?)\1\s*\)/i.exec(url_declaration);
> ```
> 
> The `(.*?)` is non-greedy and matches "anything", but it might incorrectly capture content in edge cases:
> 
> * URLs with escaped parentheses
> * URLs with escaped quotes matching the outer quote character
> 
> For example, `url("image\")other.png")` could potentially be mis-parsed.
>
> ## 4. The outer URL regex stops at the first `)` (line 55)
> ```js
> const url_declaration_regex = /url\(\s*[^)]*\)/gi;
> ```
> 
> The `[^)]*` will stop at the first `)` it encounters. This could fail with URLs containing encoded parentheses. More complex nesting could also be problematic:
> 
> ```css
> div {
> 	background: calc(url(./img.png));
> }
> ```
> 
> The regex would match `url(./img.png)` correctly here, but edge cases with encoded `)` characters in URLs could cause issues.

Thanks! This was good. I've added test cases for both and will try to figure out a solution.

> ## 5. Filename extraction doesn't account for URL-encoded characters
> ```js
> let [filename] = url.slice(VITE_ASSET_PREFIX.length).split(FRAGMENT_OR_QUERY_REGEX);
> ```
> 
> If a filename contains URL-encoded characters (e.g., `image%20name.png`), it would be compared against the `vite_assets` Set which might contain the decoded version (`image name.png`), causing a mismatch and the URL not being fixed.

Good catch. In fact, the build output is transformed to a URL encoded path, but the Vite manifest keeps the original characters, so we should definitely decode before comparing. I've added a test for it.

File: packages/kit/src/utils/css.js (hunk 16 above)
Comment (teemingc) [reply packages/kit/src/utils/css.js]: I think I've fixed the escaped characters edge case but the CSS parser is duplicating the escaped characters in the declaration value it returns. I think we need to avoid adding the same characters twice here https://github.com/sveltejs/svelte/blob/9662010a135937b7a920eaf93933be7b8b0c9c9e/packages/svelte/src/compiler/phases/1-parse/read/style.js#L508-L525 Currently, a string such as `\s` becomes `\\ss` in the returned AST

EDIT: opened a PR for it here https://github.com/sveltejs/svelte/pull/17554

EDIT2: We just need to wait for https://github.com/sveltejs/svelte/pull/17555 to be merged and update the Svelte version in Kit to get the tests passing

File: packages/kit/src/utils/css.js [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+		callback(child);
+	}
+}
+
+/**
+ * Replaces comment and string contents with whitespace.
+ * @param {string} value
+ * @returns {string}
+ */
+function tippex_comments_and_strings(value) {
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/utils/css.js]: I did have to google what tippex meant :laugh:
